### DISCUSSION TITLE
RFE: allow running ramalama from toolbox

### DISCUSSION THREAD

**Discussion by juhp** [containers/ramalama#1095] [Announcements]
Thought maybe I could use my podman flatpak-spawn hack to run ramalama from within toolbox
but it looks harder since it bind-mounts `/usr/bin/ramalama`.

Anyway this is an rfe to support toolbox: though I suppose ideally we would like the tools to be packaged in Fedora afap.

**Comment by ericcurtin** [comment]
ramalama essentially has it's own toolbox implementation. ramalama by default executes podman internally just like toolbox does.

Now one could create a ramalama toolbox if they were hell-bent on doing things via toolbox, that would be fine.

The other thing one must be wary of when using toolbox is ensuring toolbox pokes the correct holes in the toolbox container for GPU access.

Another feature I think we need sometime is some local config file, lets say someone doesn't want to type --nocontainer every time, they could do something like this:

```
$ cat ~/.ramalamaconfig
[ramalama]
	use_containers = false
        runtime = vllm
```

**Comment by rhatdan** [comment]
So this would be more about ramalama not being available on the host as silverblue.  So I install a toolbox and I want to dnf install ramalama within the toolbox.  As I understand toolbox, this should be possible and toolbox allows running of podman containers within them. so it should all work.

**Comment by ericcurtin** [comment]
Where toolbox is inconvenient right now is GPU access/AI API enablement whatever we want to call it... We should make contributions to toolbox to make that better though...

https://github.com/search?q=repo%3Acontainers%2Ftoolbox+%2Fdev%2Fkfd&type=code

"/dev/kfd" should be exposed for AMD GPUs for example... and there's no reference to it

I think we should do both... toolbox and "/usr/local" installs... Even if "/usr/local" installs are undocumented...

**Comment by debarshiray** [comment]
> Where toolbox is inconvenient right now is GPU access/AI API enablement whatever we want to call it... We should make contributions to toolbox to make that better though...

GPU access was never a problem with the free software stacks like those of AMD and Intel.  The proprietary NVIDIA stack should work from Toolbx 0.0.99.6, which is what's there on all Fedoras and CentOS Stream 9 and 10.
 
> https://github.com/search?q=repo%3Acontainers%2Ftoolbox+%2Fdev%2Fkfd&type=code

That's because Toolbx containers have the entire `/dev` from the host.

**Comment by rhatdan** [comment]
Moving this to a discussion, since this is not something I believe can be fixed in RamaLama.

**Reply by ibotty** [reply]
It depends on what you mean.  I would have hoped that lamarama could create model containers on the host from within the toolbox.  It's possible to just call out to podman on the host with `flatpak-spawn --host podman`.  I have a `podman` shell script that just did that.

Maybe that's why the following interaction with ramalama even works.

```shell
> ramalama containers
Error: no container manager (Podman, Docker) found
> ramalama --engine podman containers
WARN[0000] The deprecated BoltDB database driver is in use. This driver will be removed in the upcoming Podman 6.0 release in mid 2026. It is advised that you migrate to SQLite to avoid issues when this occurs. Set SUPPRESS_BOLTDB_WARNING environment variable to remove this message.
CONTAINER ID  IMAGE       COMMAND     CREATED     STATUS      PORTS       NAMES
```

So it is doing something.  Unfortunately starting new containers does not work, because it tries to find `llama-server`.   I thought that's not necessary when using containers?

``` shell
> ramalama --engine=podman --debug run Qwen3-Coder-Next
2026-03-07 09:49:17 - DEBUG - Checking if 8080 is available
2026-03-07 09:49:17 - DEBUG - run_cmd: podman inspect quay.io/ramalama/rocm:0.17
2026-03-07 09:49:17 - DEBUG - Working directory: None
2026-03-07 09:49:17 - DEBUG - Ignore stderr: False
2026-03-07 09:49:17 - DEBUG - Ignore all: True
2026-03-07 09:49:17 - DEBUG - env: None
2026-03-07 09:49:17 - DEBUG - Command finished with return code: 0
2026-03-07 09:49:17 - DEBUG - run_cmd: podman inspect quay.io/ramalama/rocm:0.17
2026-03-07 09:49:17 - DEBUG - Working directory: None
2026-03-07 09:49:17 - DEBUG - Ignore stderr: False
2026-03-07 09:49:17 - DEBUG - Ignore all: True
2026-03-07 09:49:17 - DEBUG - env: None
2026-03-07 09:49:17 - DEBUG - Command finished with return code: 0
2026-03-07 09:49:17 - ERROR - [Errno 2] No such file or directory: 'llama-server'
Traceback (most recent call last):
  File "/usr/lib/python3.14/site-packages/ramalama/cli.py", line 1733, in main
    args.func(args)
    ~~~~~~~~~^^^^^^
  File "/usr/lib/python3.14/site-packages/ramalama/cli.py", line 1318, in run_cli
    model.run(args, assemble_command_lazy(args))
    ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.14/site-packages/ramalama/transports/base.py", line 529, in run
    process = self.serve_nonblocking(args, cmd)
  File "/usr/lib/python3.14/site-packages/ramalama/transports/base.py", line 564, in serve_nonblocking
    process = subprocess.Popen(
        cmd,
    )
  File "/usr/lib64/python3.14/subprocess.py", line 1038, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                        pass_fds, cwd, env,
                        ^^^^^^^^^^^^^^^^^^^
    ...<5 lines>...
                        gid, gids, uid, umask,
                        ^^^^^^^^^^^^^^^^^^^^^^
                        start_new_session, process_group)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.14/subprocess.py", line 1989, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'llama-server'
Error: [Errno 2] No such file or directory: 'llama-server
```

**Comment by olliewalsh** [comment]
Implemented support for toolbox in https://github.com/containers/ramalama/pull/2871

**Comment by BansheeHero** [comment]
I think I understand the problem user is facing. I also think this might not be a problem in ramalama rather the way it is exported from the container. It should be able to call host's podman using '/usr/bin/distrobox-host-exec'

* Motivation: Run ramalama on desktops without python installed.
  * This includes: Atomic Fedora and the currently popular Bazzite
  * Gaming HW used for Bazzite leads to interest in local AI.
* The common solutions include: toolbx and distrobox
  * The access to devices (GPU) and to the filesystem is the responsibility of the tool.
* Status on my Fedora 44 installation:
  * The configuration is not respected (/etc/ramalama/ramalama.conf) (Distrobox problem)
  * The podman binary is not exposed to ramalama (TBD)

## Reproducing this on Fedora

  ```
  banshee@deck:~$ distrobox enter
  Create it now, out of image registry.fedoraproject.org/fedora-toolbox:latest? [Y/n]: y
  📦[banshee@my-distrobox ramalama]$ python -m venv ramalama && cd ramalama/ && source bin/activate
  (ramalama) 📦[banshee@my-distrobox ramalama]$ pip install ramalama
  (ramalama) 📦[banshee@my-distrobox ramalama]$ bin/ramalama -v
  ramalama version 0.23.0
  (ramalama) 📦[banshee@my-distrobox ramalama]$ whereis ramalama
  ramalama: /home/banshee/ramalama/bin/ramalama
  (ramalama) 📦[banshee@my-distrobox ramalama]$ distrobox-export --bin /home/banshee/ramalama/bin/ramalama
  /home/banshee/ramalama/bin/ramalama from my-distrobox exported successfully in /home/banshee/.local/bin.
  OK!
  (ramalama) 📦[banshee@my-distrobox ~]$ logout
  🐧banshee@deck:~$ whereis ramalama
  ramalama: /usr/bin/ramalama /etc/ramalama /usr/share/ramalama /opt/ramalama/bin/ramalama /home/banshee/.local/bin/ramalama /usr/share/man/man1/ramalama.1.gz
  🐧banshee@deck:~$ /home/banshee/.local/bin/ramalama -v
  ramalama version 0.23.0
  🐧banshee@deck:~$ /home/banshee/.local/bin/ramalama run hf://ibm-granite/granite-4.1-3b-GGUF "Say hello"
  Downloading hf://ibm-granite/granite-4.1-3b-GGUF ...
  Trying to pull hf://ibm-granite/granite-4.1-3b-GGUF ...
  Removing snapshot...
  URL pull failed and hf not available
  Error: Failed to pull model: [Errno 2] No such file or directory: '/home/banshee/.local/share/ramalama/store/huggingface/ibm-granite
  ```

### Fixing the configuration issue:

> Distrobox does forward homes, but not /etc

```
📦[banshee@my-distrobox tmp]$ mkdir /home/banshee/.config/ramalama/
📦[banshee@my-distrobox tmp]$ distrobox-host-exec cat /etc/ramalama/ramalama.conf > $HOME/.config/ramalama/ramalama.conf
📦[banshee@my-distrobox tmp]$ ramalama info | grep store 
        "store": "/mnt/sd/",
```

Now we get to the inability to call podman:

```
🐧banshee@deck:/tmp$ /home/banshee/.local/bin/ramalama run hf://ibm-granite/granite-4.1-3b-GGUF "Say hello"
Error: Failed to pull image quay.io/ramalama/ramalama:0.23 or quay.io/ramalama/ramalama:latest: [Errno 2] No such file or directory: 'podman
🐧banshee@deck:/tmp$ whereis podman
podman: /usr/bin/podman /usr/libexec/podman /usr/share/man/man1/podman.1.gz
```

### Fixing the podman issue

> Distrobox allows symlinking to break the isolation for specific binaries. 

* Workaround: `(ramalama) 📦[banshee@my-distrobox tmp]$ sudo ln -s /usr/bin/distrobox-host-exec /usr/bin/podman`
* Success: 
  
  ```
  🐧banshee@deck:/tmp$ /home/banshee/.local/bin/ramalama run hf://ibm-granite/granite-4.1-3b-GGUF "Say hello"
  af8247c8f4c9325d95ff6e2d44646eccb552811d7acf80384b2365db674e481d
  Hello! How can I help you today?
  ```

### Caveats

* I did not go through the sources to identify each time ramalama uses a subprocess call. Each call has a potential to have a host dependency broken.
* The instructions for DNF will be different.

## Moving forward

There is already progress on allowing ramalama to do this natively.
This is an alternative that works and can be boiled down to:

1. Create a distrobox
2. Install ramalama inside it.
3. Link podman with: `sudo ln -s /usr/bin/distrobox-host-exec /usr/bin/podman`
4. (Optionally) Export the binary.

**Reply by olliewalsh** [reply]
When https://github.com/containers/ramalama/pull/2871 merges, running from within a toolbox environment will work. I can investigate adding support for distrobox too in a follow-up PR.

What exactly are you trying to do though? Looks like you are installing ramalama in a distrobox environment but then trying to export it to the host? Is there a reason for doing this instead of just running within the distrobox?

**Reply by BansheeHero** [reply]
> What exactly are you trying to do though? Looks like you are installing ramalama in a distrobox environment but then trying to export it to the host? Is there a reason for doing this instead of just running within the distrobox?

Basically you cannot install python apps in Fedora Atomics without adding python to OSTree. So users that want to have these apps are encouraged to use snaps, containers, or toolboxes.
I believe this is what the OP is trying to "hack" and Bazzite users are facing when trying to use ramalama

Export is optional - that is for convenience.

**Reply by olliewalsh** [reply]
yes, indeed, that's why I added support for running inside toolbox in #2871. It works* on Fedora/Atomic.
Bazzite dropped toolbox for distrobox so I'll need to add support for that too in a follow-up PR.

* some caveats which I'll add to the docs

**Reply by BansheeHero** [reply]
> yes, indeed, that's why I added support for running inside toolbox in #2871. It works* on Fedora/Atomic. Bazzite dropped toolbox for distrobox so I'll need to add support for that too in a follow-up PR.

I noticed and that it was still open. 
I posted this because I think it is a reasonable workaround and trying to "fix" ramalama itself seemed suspicious.
The purpose of the tool like distrobox is to remove the need for apps to change.

**Reply by olliewalsh** [reply]
Yes, in general app shouldn't need to change. The podman-in-podman/docker-in-docker case is an exception.

**Reply by BansheeHero** [reply]
> When #2871 merges, running from within a toolbox environment will work. I can investigate adding support for distrobox too in a follow-up PR.
> 
> What exactly are you trying to do though? Looks like you are installing ramalama in a distrobox environment but then trying to export it to the host? Is there a reason for doing this instead of just running within the distrobox?

0.25 now generates wrong dry-run when running bench. Should I just make an issue as it is merged?
